feat: add highlight open files

nhat-vo opened the pull request:
Added highlight open file feature as described in #743. I'm not sure if there is any better way to cache this so that we do not have to loop over all buffers every time. Do you think adding to `state` is a good idea?

miversen33 commented:
I love this idea :) That said, honestly I feel like (IMO) looping over the buffers each time isn't that big of a deal, _especially_ if we only perform said change on a Buffer Change event (such as BufDelete/BufAdd event).  At the end of the day, lua is fast and a loop operation (even over several thousand items) per event loop iteration is still effectively instant. Thats not to say that thinking about "caching" or reducing the work is a bad thing, just that its **probably** not going to have a massive performance impact. Even less of an impact if we tie it only to event changes.

One thing I would be curious about, this will highlight _all_ open files with the same highlight right? Is there a way we can highlight the currently focused file _different_ if its one of the files visible in Neotree?

nhat-vo commented:
@miversen33 Thanks for your insights! I also think that caching might not result in a noticeable impact in performance, and adding the buffer list into `state` seems rather complicated so I'm not sure if I should implement this.

Regarding the focused file, are you talking about the `follow_current_file` option?

miversen33 commented:
> Regarding the focused file, are you talking about the follow_current_file option?

Yes. Does the proposed highlighting overrule that option? I am at work so I cannot test this right now lol. My main concern is that while having all open files highlighted makes a ton of sense, I wouldn't want this to supersede other options (such as `follow_current_file`)

miversen33 commented:
Reading code is hard lol. It looks like you are already using the event system for triggering the highlight logic. As long as the current focused file is distinct from the other open files, LGTM :)

cseickel commented on lua/neo-tree/sources/common/components.lua:
```diff
@@ -327,6 +327,18 @@ M.name = function(config, node, state)
   end
 
+  if config.highlight_open_files then
+    for _, buf in ipairs(vim.api.nvim_list_bufs()) do
```
This is really the wrong place to do this, because then it will do this loop for every row in the tree which is wasteful.

cseickel commented on lua/neo-tree/sources/filesystem/init.lua:
```diff
@@ -369,6 +369,17 @@ M.setup = function(config, global_config)
+      manager.subscribe(M.name, {
+        event = event,
+        handler = function()
+          manager.refresh(M.name)
```
This is the right place to collect all of the open buffers and add it to the state. Search for `modified_buffers_changed` for a good example of how to do this.

nhat-vo replied:
Ah yes I misunderstood the logic behind state and didn't manage to pass it to the update function.

nhat-vo commented on lua/neo-tree/utils.lua:
```diff
@@ -246,11 +246,13 @@ end
-    local buffer_name = vim.api.nvim_buf_get_name(buffer)
-    if buffer_name == nil or buffer_name == "" then
-      buffer_name = "[No Name]#" .. buffer
+    if vim.api.nvim_buf_is_loaded(buffer) and vim.fn.buflisted(buffer) then
```
Modified buffers should be loaded and listed, so I added this instead of creating a new method to check for open files to avoid code repetition.

nhat-vo commented on lua/neo-tree/utils.lua:
```diff
@@ -246,11 +246,13 @@ end
+      if buffer_name == nil or buffer_name == "" then
+        buffer_name = "[No Name]#" .. buffer
+      end
+      modified_buffers[buffer_name] = vim.api.nvim_buf_get_option(buffer, "modified")
```
Good idea! I'll fix it later today.

nhat-vo commented on lua/neo-tree/sources/filesystem/init.lua:
```diff
@@ -369,6 +369,15 @@ M.setup = function(config, global_config)
     })
   end
 
+  if global_config.default_component_configs.name.highlight_open_files then
```
So in that case we'll have 2 configs  (1 global config and 1 for the name component) to turn this on?

cseickel replied:
Yes, but I would default it to enabled since it's not really expensive the way it is handled. In that way, it's really one config to turn it on (the name component config) and one config to turn it off. The global config is really just fail safe for troubleshooting or people that want extreme optimization.

The alternative is to read the finalized merged config on every source and look for whether this feature is enabled, but that would be too complicated and then there's also the possibility that the data get's used for other reasons down the line.

nhat-vo commented:
@cseickel I have refactored the `modified_buffers` list to `opened_buffers` and added the configs as you suggested. I find `opened` is less ambiguous than `open` in describing the status, as `open` also refers to the action.
To make the change more uniform, I also renamed the library function in `buffers.items` to `opened` instead of `open`, in case you don't mind 😅.

cseickel reviewed (changes requested):
This looks great, there's just one more thing. Technically, the `utils` module is part of the public api and changing the signature or return type of a function in the utils module that does not start with a `_` is a breaking change that is not allowed within a major version: https://github.com/nvim-neo-tree/neo-tree.nvim#what-is-a-breaking-change

To work around that, can you just add a function that mimics the old `get_modified_buffers()`? It can be marked as deprecated using a comment:
```lua
--- DEPRECATED: This will be removed in v3
---Gets a lookup of all open buffers keyed by path with the modifed flag as the value
---@return table
M.get_modified_buffers = function()
```

Internally, it should just call the new function and convert it to the simple table of booleans that was returned before.

nhat-vo commented:
@cseickel Done!

cseickel reviewed (approved):
Looks good, thanks!
